extensions/functions_datetime.yaml

```diff
@@ -265,3 +347,278 @@ scalar_functions:
           - name: y
             value: interval_year
         return: boolean
+  -
+    name: "assume_timezone"
+    description: >-
+      Convert local timestamp to UTC-relative timestamp_tz using given local time's timezone.
+
+      Timezone strings must be as defined by IANA timezone database (https://www.iana.org/time-zones).
+      Examples: "Pacific/Marquesas", "Etc/GMT+1".
+      If timezone is not present then UTC is assumed. If timezone is invalid an error is thrown.
+    impls:
+      - args:
+          - name: x
+            value: timestamp
+          - name: timezone
+            constant: true
+            description: Timezone string from IANA tzdb.
+            value: string
+        return: timestamp_tz
+      - args:
+          - name: x
+            value: date
+          - name: timezone
+            constant: true
+            description: Timezone string from IANA tzdb.
+            value: string
+        return: timestamp_tz
+  -
+    name: "convert_timezone"
+    description: >-
+      Convert UTC-relative timestamp_tz to local timestamp using given local time's timezone.
+
+      Timezone strings must be as defined by IANA timezone database (https://www.iana.org/time-zones).
+      Examples: "Pacific/Marquesas", "Etc/GMT+1".
+      If timezone is not present then UTC is assumed. If timezone is invalid an error is thrown.
+    impls:
+      - args:
+          - name: x
+            value: timestamp_tz
+          - name: timezone
+            constant: true
+            description: Timezone string from IANA tzdb.
+            value: string
+        return: timestamp
+  -
+    name: "strptime"
+    description: >-
+      Parse string into timestamp/date/time using provided format,
+      see https://man7.org/linux/man-pages/man3/strptime.3.html for reference.
+      If timezone is present in timestamp and provided as parameter an error is thrown.
```

Comment:
How would a timezone be present in the timestamp?

Reply:
I'd go with [ISO8601](https://en.wikipedia.org/wiki/ISO_8601#Time_zone_designators), so `Z`, `UTC` or `+/-hh:mm` suffix.

Reply:
Ah.  Then how is it possible the timestamp is NOT provided as a parameter (positional arguments are always required)?

Reply:
I'd expect the `timestamp/timestamp_tz` on the output hereL `strptime(timestamp_string, ..) -> timestamp`? Sorry, I might be missing something.

Reply:
If a time zone is provided AND there is a time zone in the string value itself then wouldn't we convert?  In other words, if the user provides timezone=UTC and the timestamp string has `+07:00` then wouldn't we convert the time to UTC?

Reply:
In that case we'd convert, but in case timestamp string has no timezone info we'd interpret as provided timezone and convert to UTC or interpret as UTC and convert? I think we already have the convert in two other functions (`assume_timezone` and `convert_timezone`) and I'd prefer not to also implicitly include it here?

Reply:
So I think there are two common cases.  Let me check my understanding:

Case 1: All the values have a timezone in the string.  In this case we call the implementation that does NOT need a time zone.  This implementation will error if a string does not have a time zone.

Case 2: None of the values have a timezone in the string.  In this case we call the implementation that does need a time zone.  This implementation can error if a string has a time zone.

That just leaves the "mixed" case.  For now, I don't think we should try and solve that case in Substrait.

So, if you agree, I think we are ok here.

Reply:
Ah, sorry for a moment I thought we're discussing `strftime`.

I agree with your description of the two cases. I'm ok as is.
Perhaps we could even cut Case 1 for now as it's really `assume_timezone(strptime(x, format))` == `strptime(x, format, timezone)`?

Also from experiences in Arrow - error handling strategies are useful e.g. error on failed parsing vs return null. I'm not sure if that fits in Substrait or not.

Reply:
>  Perhaps we could even cut Case 1 for now as it's really assume_timezone(strptime(x, format)) == strptime(x, format, timezone)?

I think you mean "cut case 2"?